### PR TITLE
Changed input requirements for import/index bgen

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by tpoterba** [comment]
can't we just fix the underlying problem and give the files .bgen names?

**Comment by jigold** [comment]
No. We don't have any control over the hadoop output file naming and we don't want to do a copy-merge of the output. I had written code to rename the individual partition files, but @cseed thought accepting a directory of BGEN files was better.

**Comment by tpoterba** [comment]
The name is set in FileOutputFormat.getUniqueName

Our ByteArrayOutputFormat subclasses this, and we can trivially override it. We can add a method that takes an extension and produces a ByteArrayOutputFormat of the proper form

**Comment by tpoterba** [comment]
can also add a generator for textOutputFormat that will let us write vcf / tsv things as we like

**Comment by jigold** [comment]
I tried to do this and it wouldn't work. There was no getUniqueName to override in ByteArrayOutputFormat.

**Comment by tpoterba** [comment]
oh it's static. Hmm...

**Comment by tpoterba** [comment]
We could define a custom compression codec for .bgen and .bed that doesn't actually compress but does write the extension, haha.

don't do this

File: src/main/scala/is/hail/HailContext.scala [hunk 1]
```diff
@@ -233,14 +233,20 @@ class HailContext private(val sc: SparkContext,
     tolerance: Double = 0.2,
     nPartitions: Option[Int] = None): VariantDataset = {
 
-    val inputs = hadoopConf.globAll(files)
+    val inputs = hadoopConf.globAll(files).flatMap { file =>
+      if (hadoopConf.isDir(file))
+        hadoopConf.glob(file + "/part-*").filter(!_.getPath.getName.endsWith(".idx")).map(_.getPath.toString)
```
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: I'd filter positively for the things you're looking for, maybe something like `fs.listStatus(file).map(_.getPath.toString).filter(p => "part-[0-9]+".r.matches(p))`?

Note, there is no Regex.matches, which is super annoying.  Can you add it to utils?  I think something along the lines of:

```
class RichRegex(r: Regex) {
  def matches(s: String): Boolean = r.pattern.matcher(s).match
}
```

will work.

File: src/main/scala/is/hail/HailContext.scala [hunk 2]
```diff
@@ -233,14 +233,20 @@ class HailContext private(val sc: SparkContext,
...
+        Array(file)
+    }
 
     if (inputs.isEmpty)
       fatal(s"arguments refer to no files: '${ files.mkString(",") }'")
 
     inputs.foreach { input =>
-      if (!input.endsWith(".bgen"))
-        fatal("unknown input file type")
+      if (!input.endsWith(".bgen") && !input.contains("part")) {
```
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: I'd put this above and just warn in terms path specified by the user.  So they should give us a list of paths that end in .bgen, but can be a file or directory.

**Comment by jigold** [comment]
@cseed back to you

File: src/main/scala/is/hail/HailContext.scala [hunk 3]
```diff
@@ -233,16 +233,22 @@ class HailContext private(val sc: SparkContext,
     tolerance: Double = 0.2,
     nPartitions: Option[Int] = None): VariantDataset = {
 
-    val inputs = hadoopConf.globAll(files)
+    val inputs = hadoopConf.globAll(files).flatMap { file =>
+      if (!file.endsWith(".bgen")) {
+        warn(s"Input file does not have .bgen extension: $file")
+      }
```
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: Curly braces not necessary.

**Pull request merged**